### PR TITLE
Build: skip duplicated commands

### PR DESCRIPTION
**Pull request by humitos**
Sometimes it happens the web servers are congested and the builder retries the API call to save the command in the database. However, one of the first attempts finally ended up working resulting in the command being saved twice.

This small chunk of code checks for the existence of the command before saving it into the database and logs a warning and return 204 if it already exists, instead of re-saving it again.

It's not a perfect solution, but it could help under similar circumstances.

Related https://github.com/readthedocs/readthedocs.org/issues/10567

### CODE REVIEW

File: readthedocs/api/v2/views/model_views.py [hunk 1]
```diff
@@ -336,6 +336,11 @@ def perform_create(self, serializer):
         build_api_key = self.request.build_api_key
         if not build_api_key.project.builds.filter(pk=build_pk).exists():
             raise PermissionDenied()
+
+        if BuildCommandResult.objects.filter(**serializer.validated_data).exists():
```
Comment (stsewd) [inline readthedocs/api/v2/views/model_views.py:340]: I'd say this may be a slow query to run, maybe just limit the filtering to the build and start_time? Since we don't run commands concurrently, all of them should be guaranteed to be run sequentially (there may be an edge case with leap seconds, but that's a really really edge case :D).

**Review by stsewd — COMMENTED** [review]
We should have a test for this

File: readthedocs/api/v2/views/model_views.py (hunk 1 above)
Comment (humitos) [reply readthedocs/api/v2/views/model_views.py:340]: I don't think this query is gonna be slow as it's performed now:

```
In [12]: BuildCommandResult.objects.filter(**data).exists()
SELECT (1) AS "a"
  FROM "builds_buildcommandresult"
 WHERE ("builds_buildcommandresult"."build_id" = 7567631 AND "builds_buildcommandresult"."command" = 'git remote set-url origin https://github.com/failys/cairis.git' AND "builds_buildcommandresult"."description" = '' AND "builds_buildcommandresult"."end_time" = '2018-07-31T15:30:06.543902+00:00'::timestamptz AND "builds_buildcommandresult"."exit_code" = 0 AND "builds_buildcommandresult"."id" = 45588522 AND "builds_buildcommandresult"."output" = '' AND "builds_buildcommandresult"."start_time" = '2018-07-31T15:30:06.522766+00:00'::timestamptz)
 LIMIT 1

Execution time: 0.001359s [Database: default]
Out[12]: True
```

It takes 1.3ms in production currently.

Comment (stsewd) [reply readthedocs/api/v2/views/model_views.py:340]: My worry is more about commands with very long outputs, we don't need to check the whole output, only the build id and the start date should be unambiguous.

Comment (humitos) [reply readthedocs/api/v2/views/model_views.py:340]: Good point about long outputs 👍🏼

File: readthedocs/api/v2/views/model_views.py [hunk 2]
```diff
@@ -336,6 +336,14 @@ def perform_create(self, serializer):
         build_api_key = self.request.build_api_key
         if not build_api_key.project.builds.filter(pk=build_pk).exists():
             raise PermissionDenied()
+
+        if BuildCommandResult.objects.filter(
+            build=serializer.validated_data["build"],
+            command=serializer.validated_data["command"],
```
Comment (stsewd) [inline readthedocs/api/v2/views/model_views.py:342]: The `start_date` should be used instead of the command, since we can have the same command executed twice on purpose on the same build (like `ls` before and after a command)

Comment (humitos) [reply readthedocs/api/v2/views/model_views.py:342]: Meh, thanks. I forgot to add that field. I'll change that

File: readthedocs/api/v2/views/model_views.py [hunk 3]
```diff
@@ -336,6 +336,15 @@ def perform_create(self, serializer):
         build_api_key = self.request.build_api_key
         if not build_api_key.project.builds.filter(pk=build_pk).exists():
             raise PermissionDenied()
+
+        if BuildCommandResult.objects.filter(
+            build=serializer.validated_data["build"],
+            start_time=serializer.validated_data["start_time"],
+            command=serializer.validated_data["command"],
```
Comment (stsewd) [inline readthedocs/api/v2/views/model_views.py:343]: There is no need to filter by command, `start_time` and build are enough.

```suggestion
```

**Review by stsewd — APPROVED** [review]
